book/accessibility.md

```diff
@@ -2203,38 +2159,73 @@ class JSContext:
 
     def setAttribute(self, handle, attr, value):
         elt = self.handle_to_node[handle]
-        if attr == "role" and value == "alert" and \
-            self.getAttribute(handle, attr) != "alert":
-            self.tab.queue_alert(elt)
         elt.attributes[attr] = value
 ```
 
-Which then queues the alert if accessibility is on:
+Now we can implement the `alert` role. To do so, we'll search the
+accessiblity tree for elements with the `alert` role:
 
 ``` {.python}
 class Tab:
     def __init__(self, browser):
-        self.queued_alerts = []
+        # ...
+        self.active_alerts = []
 
-    def queue_alert(self, alert):
-        if not self.accessibility_is_on:
-            return
-        self.queued_alerts.append(alert)
-        self.set_needs_accessiblity()
+    def render(self):
+        if self.needs_accessibility:
+            # ...
+            self.active_alerts = [
+                node for node in tree_to_list(self.accessibility_tree, [])
+                if node.role == "alert"
+            ]
 ```
 
-And speaks them:
+Now, we can't just read out every `alert` at every frame; we need to
+keep track of what elements have already been read, so we don't read
+them twice:
 
 ``` {.python}
 class Tab:
-    def speak_update(self):
-        for alert in self.queued_alerts:
-            self.speak_node(alert, "New alert")
+    def __init__(self, browser):
         # ...
+        self.spoken_alerts = []
+
+    def speak_update(self):
+        for alert in self.active_alerts:
+            if alert not in self.spoken_alerts:
+                self.speak_node(alert, "New alert")
```

Comment:
perhaps add an exercise to allow setting the text content of an element in order to update the alert. Otherwise the page will have to remove the alert, wait a rAF loop, then re-add it with new innerHTML.